harper-core/src/parsers/typst.rs

```diff
@@ -0,0 +1,540 @@
+                    noun: Some(NounData {
+                        is_possessive: Some(true),
+                        ..
+                    }),
+                    ..
+                })
+            )
+        }))
+    }
+
+    #[test]
+    fn number() {
+        let source = "12 is larger than 11, but much less than 11!";
+
+        let tokens = Typst.parse_str(source);
+        let token_kinds = tokens.iter().map(|t| t.kind).collect::<Vec<_>>();
+        dbg!(&token_kinds);
+
+        assert!(matches!(
+            token_kinds.as_slice(),
+            &[
+                TokenKind::Number(OrderedFloat(12.0), None),
+                TokenKind::Space(1),
+                TokenKind::Word(_),
+                TokenKind::Space(1),
+                TokenKind::Word(_),
+                TokenKind::Space(1),
+                TokenKind::Word(_),
+                TokenKind::Space(1),
+                TokenKind::Number(OrderedFloat(11.0), None),
+                TokenKind::Punctuation(Punctuation::Comma),
+                TokenKind::Space(1),
+                TokenKind::Word(_),
+                TokenKind::Space(1),
+                TokenKind::Word(_),
+                TokenKind::Space(1),
+                TokenKind::Word(_),
+                TokenKind::Space(1),
+                TokenKind::Word(_),
+                TokenKind::Space(1),
+                TokenKind::Number(OrderedFloat(11.0), None),
+                TokenKind::Punctuation(Punctuation::Bang),
+            ]
+        ))
+    }
+
+    #[test]
+    fn math_unlintable() {
+        let source = "$12 > 11$, $12 << 11!$";
+
+        let tokens = Typst.parse_str(source);
+        let token_kinds = tokens.iter().map(|t| t.kind).collect::<Vec<_>>();
+        dbg!(&token_kinds);
+
+        assert!(matches!(
+            token_kinds.as_slice(),
+            &[
+                TokenKind::Unlintable,
+                TokenKind::Punctuation(Punctuation::Comma),
+                TokenKind::Space(1),
+                TokenKind::Unlintable,
+            ]
+        ))
+    }
+
+    #[test]
+    fn dict_parsing() {
+        let source = r#"#let dict = (
+                        name: "Typst",
+                        born: 2019,
+                      )"#;
+
+        let tokens = Typst.parse_str(source);
+        let token_kinds = tokens.iter().map(|t| t.kind).collect::<Vec<_>>();
+        dbg!(&token_kinds);
+
+        let charslice = source.chars().collect_vec();
+        assert_eq!(tokens[3].span.get_content_string(&charslice), "Typst");
+        assert!(matches!(
+            token_kinds.as_slice(),
+            &[
+                TokenKind::Word(_),
+                TokenKind::Word(_),
+                TokenKind::Punctuation(Punctuation::Quote { .. }),
+                TokenKind::Word(_),
+                TokenKind::Punctuation(Punctuation::Quote { .. }),
+                TokenKind::Word(_),
+                TokenKind::Number(OrderedFloat(2019.0), None),
+            ]
+        ))
+    }
```

Comment:
Going from intuition, the quotes should not count as actual quotes, because it's just how you construct a string in Typst.

The second problem is the question of whether the custom var/func names should be checked. Ideally this should be configurable.

On one hand, it would be nice to check the correctness of the variable names etc., but most IDEs already can do that and you can't use commas and periods, for example, as well as the name can be lowercase or uppercase. So I don't see any good reason to count those in by default.

Plus, it will mess up the sentence structure and will give false positives (like "the word should start with a capital letter" while it's a variable name). So the dictionary names definitely should not by checked (by default).

Going deeper, the problem with Typst is that you can construct text from code mode like `#range(5).map(_ => "a")` which will result in "aaaaa". There is a `lorem()` function that can output a ton of Latin text, and so can do user functions. This is probably an edge case that is unsolvable, unless you get the resolved AST or string, meaning all the code part is gone and the raw text (maybe with some styling) is left. I have yet to learn all the insides of Typst, especially since several previous rewrites (of some parts). So I can't be sure about the Rust's side of Typst in this PR.

In general, maybe it's not a great idea to check anything in code mode, only in markup. But I think this can be resolved if harper supports isolated text. So if you have 3 plain text paragraphs (and nothing else), then it would count as 1 isolated text chunk. If you also have some separate text, e.g., an image caption, then it would be a second text chunk. And since these chunks are isolated, you won't get any false positives from checking them together.

In a table with many cells you would have a separate text block for each cell, because they can have different content that isn't a sentence. If this is possible, then you can also safely check dictionary value etc.

Reply:
> On one hand, it would be nice to check the correctness of the variable names etc., but most IDEs already can do that and you can't use commas and periods, for example, as well as the name can be lowercase or uppercase. So I don't see any good reason to count those in by default.
>
> Plus, it will mess up the sentence structure and will give false positives (like "the word should start with a capital letter" while it's a variable name). So the dictionary names definitely should not by checked (by default).

Harper has some degree of support for identifiers already for code comments, I don't think it would be too complicated to extend that here, we'd just need to extract all the identifiers to build the file dictionary.

Reply:
> Going deeper, the problem with Typst is that you can construct text from code mode like #range(5).map(_ => "a") which will result in "aaaaa". There is a lorem() function that can output a ton of Latin text, and so can do user functions. This is probably an edge case that is unsolvable, unless you get the resolved AST or string, meaning all the code part is gone and the raw text (maybe with some styling) is left. I have yet to learn all the insides of Typst, especially since several previous rewrites (of some parts). So I can't be sure about the Rust's side of Typst in this PR.

Yes, this whole thing was the biggest issue when I was trying to support LaTeX as well. I agree that it is probably unsolvable. I think we could theoretically expand things using typst's libraries, but I assume it would have horrible performance.